
 # Dealing With Null Values Solution Notebook.
 <br>
 <br>
 It's time to get your hands dirty dude!!


# Import Libraries

In [ ]:
import numpy as np
import pandas as pd

 Python list <b>headers</b> containing name of headers


In [ ]:
# These are the headers which you need to link to your dataset.
headers = ["symboling","normalized-losses","make","fuel-type","aspiration", "num-of-doors","body-style",
         "drive-wheels","engine-location","wheel-base", "length","width","height","curb-weight","engine-type",
         "num-of-cylinders", "engine-size","fuel-system","bore","stroke","compression-ratio","horsepower",
         "peak-rpm","city-mpg","highway-mpg","price"]

# Import Datasets

Use the Pandas method <b>read_csv()</b> to load the data from the web address. Set the parameter  "names" equal to the Python list "headers".


In [ ]:
df = pd.read_csv('car_dataset.data', names = headers)

 Use the method <b>head()</b> to display the first five rows of the dataframe.


In [ ]:
# To see what the data set looks like, we'll use the head() method.
df.head(15)


symboling normalized-losses         make fuel-type aspiration  \
0           3                 ?  alfa-romero       gas        std   
1           3                 ?  alfa-romero       gas        std   
2           1                 ?  alfa-romero       gas        std   
3           2               164         audi       gas        std   
4           2               164         audi       gas        std   
5           2                 ?         audi       gas        std   
6           1               158         audi       gas        std   
7           1                 ?         audi       gas        std   
8           1               158         audi       gas      turbo   
9           0                 ?         audi       gas      turbo   
10          2               192          bmw       gas        std   
11          0               192          bmw       gas        std   
12          0               188          bmw       gas        std   
13          0               188          bmw       gas        std   
14          1                 ?          bmw       gas        std   

   num-of-doors   body-style drive-wheels engine-location  wheel-base  ...  \
0           two  convertible          rwd           front        88.6  ...   
1           two  convertible          rwd           front        88.6  ...   
2           two    hatchback          rwd           front        94.5  ...   
3          four        sedan          fwd           front        99.8  ...   
4          four        sedan          4wd           front        99.4  ...   
5           two        sedan          fwd           front        99.8  ...   
6          four        sedan          fwd           front       105.8  ...   
7          four        wagon          fwd           front       105.8  ...   
8          four        sedan          fwd           front       105.8  ...   
9           two    hatchback          4wd           front        99.5  ...   
10          two        sedan          rwd           front       101.2  ...   
11         four        sedan          rwd           front       101.2  ...   
12          two        sedan          rwd           front       101.2  ...   
13         four        sedan          rwd           front       101.2  ...   
14         four        sedan          rwd           front       103.5  ...   

    engine-size  fuel-system  bore  stroke compression-ratio horsepower  \
0           130         mpfi  3.47    2.68               9.0        111   
1           130         mpfi  3.47    2.68               9.0        111   
2           152         mpfi  2.68    3.47               9.0        154   
3           109         mpfi  3.19    3.40              10.0        102   
4           136         mpfi  3.19    3.40               8.0        115   
5           136         mpfi  3.19    3.40               8.5        110   
6           136         mpfi  3.19    3.40               8.5        110   
7           136         mpfi  3.19    3.40               8.5        110   
8           131         mpfi  3.13    3.40               8.3        140   
9           131         mpfi  3.13    3.40               7.0        160   
10          108         mpfi  3.50    2.80               8.8        101   
11          108         mpfi  3.50    2.80               8.8        101   
12          164         mpfi  3.31    3.19               9.0        121   
13          164         mpfi  3.31    3.19               9.0        121   
14          164         mpfi  3.31    3.19               9.0        121   

    peak-rpm city-mpg highway-mpg  price  
0       5000       21          27  13495  
1       5000       21          27  16500  
2       5000       19          26  16500  
3       5500       24          30  13950  
4       5500       18          22  17450  
5       5500       19          25  15250  
6       5500       19          25  17710  
7       5500       19          25  18920  
8       5500       17          20  23875  
9       5500       

As we can see, several question marks appeared in the dataframe; those are missing values which may hinder our further analysis.

<div>So, how do we identify all those missing values and deal with them?</div>

<b>How to work with missing data?</b>

Steps for working with missing data:

<ol>
    <li>Identify missing data</li>
    <li>deal with missing data</li>
    <li>correct data format</li>
</ol>


<h1 id="identify_handle_missing_values">Identify and handle missing values</h1>

<h2 id="identify_missing_values">Identify missing values</h2>
<h3>Convert "?" to NaN</h3>
In the car dataset, missing data comes with the question mark "?".
We replace "?" with NaN (Not a Number), which is Python's default missing value marker, for reasons of computational speed and convenience. Here we use the function:
 <pre>.replace(A, B, inplace = True) </pre>
to replace A by B


In [ ]:


# replace "?" to NaN
df.replace('?',np.nan,inplace=True)

In [ ]:
df.head()

symboling normalized-losses         make fuel-type aspiration num-of-doors  \
0          3               NaN  alfa-romero       gas        std          two   
1          3               NaN  alfa-romero       gas        std          two   
2          1               NaN  alfa-romero       gas        std          two   
3          2               164         audi       gas        std         four   
4          2               164         audi       gas        std         four   

    body-style drive-wheels engine-location  wheel-base  ...  engine-size  \
0  convertible          rwd           front        88.6  ...          130   
1  convertible          rwd           front        88.6  ...          130   
2    hatchback          rwd           front        94.5  ...          152   
3        sedan          fwd           front        99.8  ...          109   
4        sedan          4wd           front        99.4  ...          136   

   fuel-system  bore  stroke compression-ratio horsepower  peak-rpm city-mpg  \
0         mpfi  3.47    2.68               9.0        111      5000       21   
1         mpfi  3.47    2.68               9.0        111      5000       21   
2         mpfi  2.68    3.47               9.0        154      5000       19   
3         mpfi  3.19    3.40              10.0        102      5500       24   
4         mpfi  3.19    3.40               8.0        115      5500       18   

  highway-mpg  price  
0          27  13495  
1          27  16500  
2          26  16500  
3          30  13950  
4          22  17450  

[5 rows x 26 columns]

Identify_missing_values

<h2>Evaluating for Missing Data</h2>

The missing values are converted to default. We use the following functions to identify these missing values. There are two methods to detect missing data:

<ol>
    <li><b>.isnull()</b></li>
    <li><b>.notnull()</b></li>
</ol>
The output is a boolean value indicating whether the value that is passed into the argument is in fact missing data.


In [ ]:
missing_data = df.isnull()
missing_data.head(10)

symboling  normalized-losses   make  fuel-type  aspiration  num-of-doors  \
0      False               True  False      False       False         False   
1      False               True  False      False       False         False   
2      False               True  False      False       False         False   
3      False              False  False      False       False         False   
4      False              False  False      False       False         False   
5      False               True  False      False       False         False   
6      False              False  False      False       False         False   
7      False               True  False      False       False         False   
8      False              False  False      False       False         False   
9      False               True  False      False       False         False   

   body-style  drive-wheels  engine-location  wheel-base  ...  engine-size  \
0       False         False            False       False  ...        False   
1       False         False            False       False  ...        False   
2       False         False            False       False  ...        False   
3       False         False            False       False  ...        False   
4       False         False            False       False  ...        False   
5       False         False            False       False  ...        False   
6       False         False            False       False  ...        False   
7       False         False            False       False  ...        False   
8       False         False            False       False  ...        False   
9       False         False            False       False  ...        False   

   fuel-system   bore  stroke  compression-ratio  horsepower  peak-rpm  \
0        False  False   False              False       False     False   
1        False  False   False              False       False     False   
2        False  False   False              False       False     False   
3        False  False   False              False       False     False   
4        False  False   False              False       False     False   
5        False  False   False              False       False     False   
6        False  False   False              False       False     False   
7        False  False   False              False       False     False   
8        False  False   False              False       False     False   
9        False  False   False              False       False     False   

   city-mpg  highway-mpg  price  
0     False        False  False  
1     False        False  False  
2     False        False  False  
3     False        False  False  
4     False        False  False  
5     False        False  False  
6     False        False  False  
7     False        False  False  
8     False        False  False  
9     False        False   True  

[10 rows x 26 columns]

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 205 entries, 0 to 204
Data columns (total 26 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   symboling          205 non-null    int64  
 1   normalized-losses  164 non-null    object 
 2   make               205 non-null    object 
 3   fuel-type          205 non-null    object 
 4   aspiration         205 non-null    object 
 5   num-of-doors       203 non-null    object 
 6   body-style         205 non-null    object 
 7   drive-wheels       205 non-null    object 
 8   engine-location    205 non-null    object 
 9   wheel-base         205 non-null    float64
 10  length             205 non-null    float64
 11  width              205 non-null    float64
 12  height             205 non-null    float64
 13  curb-weight        205 non-null    int64  
 14  engine-type        205 non-null    object 
 15  num-of-cylinders   205 non-null    object 
 16  engine-size        205 non

"True" stands for missing value, while "False" stands for not missing value.


<h2>Count missing values in each column</h2>
<p>
Using a for loop in Python, we can quickly figure out the number of missing values in each column. As mentioned above, "True" represents a missing value, "False"  means the value is present in the dataset.  In the body of the for loop the method  ".value_counts()"  counts the number of "True" values.
</p>


In [ ]:
for column in headers:
  print(column)
  print(missing_data[column].value_counts())
  print("  ")

symboling
False    205
Name: symboling, dtype: int64
  
normalized-losses
False    164
True      41
Name: normalized-losses, dtype: int64
  
make
False    205
Name: make, dtype: int64
  
fuel-type
False    205
Name: fuel-type, dtype: int64
  
aspiration
False    205
Name: aspiration, dtype: int64
  
num-of-doors
False    203
True       2
Name: num-of-doors, dtype: int64
  
body-style
False    205
Name: body-style, dtype: int64
  
drive-wheels
False    205
Name: drive-wheels, dtype: int64
  
engine-location
False    205
Name: engine-location, dtype: int64
  
wheel-base
False    205
Name: wheel-base, dtype: int64
  
length
False    205
Name: length, dtype: int64
  
width
False    205
Name: width, dtype: int64
  
height
False    205
Name: height, dtype: int64
  
curb-weight
False    205
Name: curb-weight, dtype: int64
  
engine-type
False    205
Name: engine-type, dtype: int64
  
num-of-cylinders
False    205
Name: num-of-cylinders, dtype: int64
  
engine-size
False    205
Name: engine-si

Based on the summary above, each column has 205 rows of data, seven columns containing missing data:

<ol>
    <li>"normalized-losses": 41 missing data</li>
    <li>"num-of-doors": 2 missing data</li>
    <li>"bore": 4 missing data</li>
    <li>"stroke" : 4 missing data</li>
    <li>"horsepower": 2 missing data</li>
    <li>"peak-rpm": 2 missing data</li>
    <li>"price": 4 missing data</li>
</ol>


<h2 id="deal_missing_values">Deal with missing data</h2>
<br>
<h3>How to deal with missing data?</h3>

<ol>
    <li>drop data<br>
        a. drop the whole row<br>
        b. drop the whole column
    </li>
    <li>replace data<br>
        a. replace it by mean<br>
        b. replace it by frequency<br>
    </li>
</ol>


Whole columns should be dropped only if most entries in the column are empty. In our dataset, none of the columns are empty enough to drop entirely.
We have some freedom in choosing which method to replace data; however, some methods may seem more reasonable than others. We will apply each method to many different columns:

<b>Replace by mean:</b>

<ul>
    <li>"normalized-losses": 41 missing data, replace them with mean</li>
    <li>"stroke": 4 missing data, replace them with mean</li>
    <li>"bore": 4 missing data, replace them with mean</li>
    <li>"horsepower": 2 missing data, replace them with mean</li>
    <li>"peak-rpm": 2 missing data, replace them with mean</li>
</ul>

<b>Replace by frequency:</b>

<ul>
    <li>"num-of-doors": 2 missing data, replace them with "four".
        <ul>
            <li>Reason: 84% sedans is four doors. Since four doors is most frequent, it is most likely to occur</li>
        </ul>
    </li>
</ul>

<b>Drop the whole row:</b>

<ul>
    <li>"price": 4 missing data, simply delete the whole row
        <ul>
            <li>Reason: price is what we want to predict. Any data entry without price data cannot be used for prediction; therefore any row now without price data is not useful to us</li>
        </ul>
    </li>
</ul>


<h3>Calculate the mean of the normalized-losses column </h3>


In [ ]:
avg_norm_losses = df['normalized-losses'].astype("float").mean()
avg_norm_losses

122.0

<h3>Replace "NaN" by mean value in "normalized-losses" column</h3>


In [ ]:
df['normalized-losses'].replace(np.nan,avg_norm_losses,inplace=True)
df.head()

symboling normalized-losses         make fuel-type aspiration num-of-doors  \
0          3             122.0  alfa-romero       gas        std          two   
1          3             122.0  alfa-romero       gas        std          two   
2          1             122.0  alfa-romero       gas        std          two   
3          2               164         audi       gas        std         four   
4          2               164         audi       gas        std         four   

    body-style drive-wheels engine-location  wheel-base  ...  engine-size  \
0  convertible          rwd           front        88.6  ...          130   
1  convertible          rwd           front        88.6  ...          130   
2    hatchback          rwd           front        94.5  ...          152   
3        sedan          fwd           front        99.8  ...          109   
4        sedan          4wd           front        99.4  ...          136   

   fuel-system  bore  stroke compression-ratio horsepower  peak-rpm city-mpg  \
0         mpfi  3.47    2.68               9.0        111      5000       21   
1         mpfi  3.47    2.68               9.0        111      5000       21   
2         mpfi  2.68    3.47               9.0        154      5000       19   
3         mpfi  3.19    3.40              10.0        102      5500       24   
4         mpfi  3.19    3.40               8.0        115      5500       18   

  highway-mpg  price  
0          27  13495  
1          27  16500  
2          26  16500  
3          30  13950  
4          22  17450  

[5 rows x 26 columns]

<h3>Calculate the mean value for 'bore' column</h3>


In [ ]:
avg_bore = df['bore'].astype('float').mean()
avg_bore

3.3297512437810943

<h3>Replace NaN by mean value for bore column</h3>


In [ ]:
df['bore'].replace(np.nan,avg_bore,inplace=True)

<h3>Calculate the mean value for 'stroke' column</h3>

In [ ]:
# Write your code below and press Shift+Enter to execute
avg_stroke = df['stroke'].astype('float').mean()


<h3>Replace NaN by mean value for stroke column</h3>

In [ ]:
df['stroke'].replace(np.nan,avg_stroke,inplace=True)

<details><summary>Click here for the solution</summary>

```python
#Calculate the mean vaule for "stroke" column
avg_stroke = df["stroke"].astype("float").mean(axis = 0)
print("Average of stroke:", avg_stroke)

# replace NaN by mean value in "stroke" column
df["stroke"].replace(np.nan, avg_stroke, inplace = True)
```

</details>


<h4>Calculate the mean value for the  'horsepower' column:</h4>


In [ ]:
avg_horsepower  = df['horsepower'].astype('float').mean()


<h4>Replace "NaN" by mean value:</h4>


In [ ]:
df['horsepower'].replace(np.nan,avg_horsepower,inplace=True)

<h4>Calculate the mean value for 'peak-rpm' column:</h4>


In [ ]:
avg_peak = df['peak-rpm'].astype('float').mean()


<h4>Replace NaN by mean value:</h4>


In [ ]:
df['peak-rpm'].replace(np.nan,avg_peak,inplace=True)

In [ ]:
avg_peak

5125.369458128079

In [ ]:
df.head()

symboling normalized-losses         make fuel-type aspiration num-of-doors  \
0          3             122.0  alfa-romero       gas        std          two   
1          3             122.0  alfa-romero       gas        std          two   
2          1             122.0  alfa-romero       gas        std          two   
3          2               164         audi       gas        std         four   
4          2               164         audi       gas        std         four   

    body-style drive-wheels engine-location  wheel-base  ...  engine-size  \
0  convertible          rwd           front        88.6  ...          130   
1  convertible          rwd           front        88.6  ...          130   
2    hatchback          rwd           front        94.5  ...          152   
3        sedan          fwd           front        99.8  ...          109   
4        sedan          4wd           front        99.4  ...          136   

   fuel-system  bore  stroke compression-ratio horsepower  peak-rpm city-mpg  \
0         mpfi  3.47    2.68               9.0        111      5000       21   
1         mpfi  3.47    2.68               9.0        111      5000       21   
2         mpfi  2.68    3.47               9.0        154      5000       19   
3         mpfi  3.19    3.40              10.0        102      5500       24   
4         mpfi  3.19    3.40               8.0        115      5500       18   

  highway-mpg  price  
0          27  13495  
1          27  16500  
2          26  16500  
3          30  13950  
4          22  17450  

[5 rows x 26 columns]

<br>

## Alright you have done great job till now, but its not finished yet dude!!!
<br>

<ul>
<li>Now we have done for replacing with mean.</li>
<li>Its time to replace null values with most frequent values </li>
<li>We will do this for num-of-doors column</li>
</ul>

<h4>Find the most frequent value for num of doors</h4>

In [ ]:
df['num-of-doors'].value_counts()

four    114
two      89
Name: num-of-doors, dtype: int64

<h4>We can see that four doors are the most common type. We can also use the ".idxmax()" method to calculate for us the most common type automatically:</h4>


In [ ]:
df['num-of-doors'].value_counts().idxmax()

'four'

<h4>The replacement procedure is very similar to what we have seen previously</h4>


In [ ]:
#replace the missing 'num-of-doors' values by the most frequent
df['num-of-doors'].replace(np.nan,'four',inplace=True)

## Finally, let's drop all rows that do not have price data:


In [ ]:
# simply drop whole row with NaN in "price" column
df.dropna(subset='price',axis=0,inplace=True)

# reset index, because we droped two rows
df.reset_index(drop=True,inplace=True)

In [ ]:
df.head(50)

symboling normalized-losses         make fuel-type aspiration  \
0           3             122.0  alfa-romero       gas        std   
1           3             122.0  alfa-romero       gas        std   
2           1             122.0  alfa-romero       gas        std   
3           2               164         audi       gas        std   
4           2               164         audi       gas        std   
5           2             122.0         audi       gas        std   
6           1               158         audi       gas        std   
7           1             122.0         audi       gas        std   
8           1               158         audi       gas      turbo   
9           2               192          bmw       gas        std   
10          0               192          bmw       gas        std   
11          0               188          bmw       gas        std   
12          0               188          bmw       gas        std   
13          1             122.0          bmw       gas        std   
14          0             122.0          bmw       gas        std   
15          0             122.0          bmw       gas        std   
16          0             122.0          bmw       gas        std   
17          2               121    chevrolet       gas        std   
18          1                98    chevrolet       gas        std   
19          0                81    chevrolet       gas        std   
20          1               118        dodge       gas        std   
21          1               118        dodge       gas        std   
22          1               118        dodge       gas      turbo   
23          1               148        dodge       gas        std   
24          1               148        dodge       gas        std   
25          1               148        dodge       gas        std   
26          1               148        dodge       gas      turbo   
27         -1               110        dodge       gas        std   
28          3               145        dodge       gas      turbo   
29          2               137        honda       gas        std   
30          2               137        honda       gas        std   
31          1               101        honda       gas        std   
32          1               101        honda       gas        std   
33          1               101        honda       gas        std   
34          0               110        honda       gas        std   
35          0                78        honda       gas        std   
36          0               106        honda       gas        std   
37          0               106        honda       gas        std   
38          0                85        honda       gas        std   
39          0                85        honda       gas        std   
40          0                85        honda       gas        std   
41          1               107        honda       gas        std   
42          0             122.0        isuzu       gas        std   
43          2             122.0        isuzu       gas        std   
44          0               145       jaguar       gas        std   
45          0             122.0       jaguar       gas        std   
46          0             122.0       jaguar       gas        std   
47          1               104        mazda       gas        std   
48          1               104        mazda       gas        std   
49          1               104        mazda       gas        std   

   num-of-doors   body-style drive-wheels engine-location  wheel-base  ...  \
0           two  convertible          rwd           front        88.6  ...   
1           two  convertible          rwd           front        88.6  ...   
2           two    hatchback          rwd           front        94.5  ...   
3          four        sedan          fwd           front        99.8  ...   
4          four        sedan          4wd           front        99.4  ...   
5           two 

<h2> Lets Check It</h2>

In [ ]:
missing_data = df.isnull()


for column in headers:
  print(column)
  print(missing_data[column].value_counts())
  print("  ")

symboling
False    201
Name: symboling, dtype: int64
  
normalized-losses
False    201
Name: normalized-losses, dtype: int64
  
make
False    201
Name: make, dtype: int64
  
fuel-type
False    201
Name: fuel-type, dtype: int64
  
aspiration
False    201
Name: aspiration, dtype: int64
  
num-of-doors
False    201
Name: num-of-doors, dtype: int64
  
body-style
False    201
Name: body-style, dtype: int64
  
drive-wheels
False    201
Name: drive-wheels, dtype: int64
  
engine-location
False    201
Name: engine-location, dtype: int64
  
wheel-base
False    201
Name: wheel-base, dtype: int64
  
length
False    201
Name: length, dtype: int64
  
width
False    201
Name: width, dtype: int64
  
height
False    201
Name: height, dtype: int64
  
curb-weight
False    201
Name: curb-weight, dtype: int64
  
engine-type
False    201
Name: engine-type, dtype: int64
  
num-of-cylinders
False    201
Name: num-of-cylinders, dtype: int64
  
engine-size
False    201
Name: engine-size, dtype: int64
  
fuel-s

Hurrey!!!
Now we have our dataset with no missing values.
<br>
Isn't It?

In [ ]:
df.to_csv("dealingmissingvals.csv")

<br>
<br>

# REALLY IT WAS A GREAT JOB DUDE!!!
<h4>LETS START LEARNING DATA FORMATING & DATA STANDARDIZATION.</h4>
<h4>There you will learn how to correct the data types of each columns in proper way. </h4>
<br>
<br>

